packages/go_router/lib/src/navigator_builder.dart

```diff
@@ -0,0 +1,124 @@
+import 'package:flutter/cupertino.dart';
+
+import '../go_router.dart';
+import 'builder.dart';
+import 'configuration.dart';
+import 'match.dart';
+import 'matching.dart';
+import 'parser.dart';
+
+/// Provides support for building Navigators for routes.
+class RouteNavigatorBuilder extends ShellNavigatorBuilder {
+  /// Constructs a NavigatorBuilder.
+  RouteNavigatorBuilder(
+    this.routeBuilder,
+    this.state,
+    this.currentRoute,
+    this.heroController,
+    this.navigatorKeyForCurrentRoute,
+    this.pages,
+    this.pagePopContext,
+  );
+
+  /// The route builder.
+  final RouteBuilder routeBuilder;
+
+  @override
+  final GoRouterState state;
+
+  @override
+  final ShellRouteBase currentRoute;
+
+  /// The hero controller.
+  final HeroController heroController;
+
+  @override
+  final GlobalKey<NavigatorState> navigatorKeyForCurrentRoute;
+
+  /// The pages for the current route.
+  final List<Page<Object?>> pages;
+
+  /// The page pop context.
+  final PagePopContext pagePopContext;
+
+  /// Builds a navigator.
+  static Widget buildNavigator(
+    PopPageCallback onPopPage,
+    List<Page<Object?>> pages,
+    Key? navigatorKey, {
+    List<NavigatorObserver>? observers,
+    String? restorationScopeId,
+    HeroController? heroController,
+  }) {
+    final Widget navigator = Navigator(
+      key: navigatorKey,
+      restorationScopeId: restorationScopeId,
+      pages: pages,
+      observers: observers ?? const <NavigatorObserver>[],
+      onPopPage: onPopPage,
+    );
+    if (heroController != null) {
+      return HeroControllerScope(
+        controller: heroController,
+        child: navigator,
+      );
+    } else {
+      return navigator;
+    }
+  }
+
+  @override
+  Widget buildNavigatorForCurrentRoute({
+    List<NavigatorObserver>? observers,
+    String? restorationScopeId,
+    GlobalKey<NavigatorState>? navigatorKey,
+  }) {
+    return buildNavigator(
+      pagePopContext.onPopPage,
+      pages,
+      navigatorKey ?? navigatorKeyForCurrentRoute,
+      observers: observers,
+      restorationScopeId: restorationScopeId,
+      heroController: heroController,
+    );
+  }
+
+  @override
+  Future<Widget> buildPreloadedShellNavigator({
+    required BuildContext context,
+    required String location,
+    Object? extra,
+    required GlobalKey<NavigatorState> navigatorKey,
+    required ShellRouteBase parentShellRoute,
+    List<NavigatorObserver>? observers,
+    String? restorationScopeId,
+  }) {
+    // Parse a RouteMatchList from location and handle any redirects
+    final GoRouteInformationParser parser =
+        GoRouter.of(context).routeInformationParser;
+    final Future<RouteMatchList> routeMatchList =
+        parser.parseRouteInformationWithDependencies(
```

Comment:
This is actually the reason GoRouteInformationParser is used here. And there is an assert further down validating that the potentially redirected location still points to a location below the current StatefulShellRoute (although it could probably use a message in the assert). There should however probably also be an assert that validates that the new location also belongs to the correct branch.

Reply:
> there is an assert further down validating that the potentially redirected location still points to a location below the current StatefulShellRoute

What if it doesn't? I think it is valid that the redirection take it to a different place.

Reply:
I guess that's true. And if so, at that point preloading should just be aborted I guess (instead of failing on assert).

Reply:
I think preload should not go through redirect. It doesn't make sense even if the redirection only bring you to a different branch under the same statefulShellRoute. For example: you want to preload branch A, but it preload branchB instead because of the redirection.

Reply:
How bad is performance if there is no preload. It seems to me this will be a source of bugs or even a footgun when developers might just blindly slap preload= true on every branches